fedimint-core/src/db/mod.rs

```diff
@@ -2070,3 +2196,138 @@ mod tests {
+            Some(TestVal(2)),
+            "should notify"
+        );
+    }
+
+    #[tokio::test]
+    async fn test_wait_key_before_insert() {
+        let key = TestKey(1);
+        let val = TestVal(2);
+        let db = Database::new(MemDatabase::new(), ModuleDecoderRegistry::default());
+
+        let mut tx = db.begin_transaction().await;
+        let key_task = waiter(&db, TestKey(1)).await;
+        tx.insert_new_entry(&key, &val).await.expect("DB error");
+        tx.commit_tx().await.expect("DB error");
+
+        assert_eq!(
+            future_returns_shortly(async { key_task.await.unwrap() }).await,
+            Some(TestVal(2)),
+            "should notify"
+        );
+    }
+
+    #[tokio::test]
+    async fn test_wait_key_after_insert() {
+        let key = TestKey(1);
+        let val = TestVal(2);
+        let db = Database::new(MemDatabase::new(), ModuleDecoderRegistry::default());
+
+        let mut tx = db.begin_transaction().await;
+        tx.insert_new_entry(&key, &val).await.expect("DB error");
+
+        let key_task = waiter(&db, TestKey(1)).await;
+
+        tx.commit_tx().await.expect("DB error");
+
+        assert_eq!(
+            future_returns_shortly(async { key_task.await.unwrap() }).await,
+            Some(TestVal(2)),
+            "should notify"
+        );
+    }
+
+    #[tokio::test]
+    async fn test_wait_key_after_notify() {
+        let key = TestKey(1);
+        let val = TestVal(2);
+        let db = Database::new(MemDatabase::new(), ModuleDecoderRegistry::default());
+
+        let mut tx = db.begin_transaction().await;
+        tx.insert_new_entry(&key, &val).await.expect("DB error");
+
+        let key_task = waiter(&db, TestKey(1)).await;
+        tx.commit_tx().await.expect("DB error");
+
+        assert_eq!(
+            future_returns_shortly(async { key_task.await.unwrap() }).await,
+            Some(TestVal(2)),
+            "should notify"
+        );
+    }
+
+    #[tokio::test]
+    async fn test_wait_key_after_commit() {
+        let key = TestKey(1);
+        let val = TestVal(2);
+        let db = Database::new(MemDatabase::new(), ModuleDecoderRegistry::default());
+
+        let mut tx = db.begin_transaction().await;
+        tx.insert_new_entry(&key, &val).await.expect("DB error");
+        tx.commit_tx().await.expect("DB error");
+
+        let key_task = waiter(&db, TestKey(1)).await;
+        assert_eq!(
+            future_returns_shortly(async { key_task.await.unwrap() }).await,
+            Some(TestVal(2)),
+            "should notify"
+        );
+    }
+
+    #[tokio::test]
+    async fn test_wait_key_no_transaction() {
+        let db = Database::new(MemDatabase::new(), ModuleDecoderRegistry::default());
+
+        let key_task = waiter(&db, TestKey(1)).await;
+        assert_eq!(
+            future_returns_shortly(async { key_task.await.unwrap() }).await,
+            None,
+            "should not notify"
+        );
+    }
```

Comment:
Can we add a test that tests using an IsolatedDatabase? (i.e `db.new_isolated(<module_instance_id>)`)